sale_order_line_delivery_state/tests/test_sale_line_delivery_state.py

```diff
@@ -0,0 +1,116 @@
+                        0,
+                        {
+                            "product_id": product.id,
+                            "product_uom": cls.uom.id,
+                            "product_uom_qty": 3,
+                            "price_unit": 2950.00,
+                            "name": product.name,
+                        },
+                    )
+                ],
+            }
+        )
+
+    def _mock_delivery(self, delivery_prod=None):
+        delivery_prod = delivery_prod or self.delivery_cost
+        return mock.patch.object(
+            type(self.env["sale.order.line"]),
+            "_is_delivery",
+            lambda self: self.product_id == delivery_prod,
+        )
+
+    def _add_delivery_cost_line(self):
+        self.env["sale.order.line"].create(
+            {
+                "order_id": self.order.id,
+                "name": "Delivery cost",
+                "product_id": self.delivery_cost.id,
+                "product_uom_qty": 1,
+                "product_uom": self.uom.id,
+                "price_unit": 10.0,
+            }
+        )
+
+    def test_no_delivery(self):
+        self.assertEqual(self.order.order_line[0].delivery_state, "no")
+
+    def test_unprocessed_delivery(self):
+        self.order.action_confirm()
+        self.assertEqual(self.order.order_line[0].delivery_state, "unprocessed")
+
+    def test_partially(self):
+        self.order.action_confirm()
+        self.order.order_line[0].qty_delivered = 2
+        self.assertEqual(self.order.order_line[0].delivery_state, "partially")
+
+    def test_delivery_done(self):
+        self.order.action_confirm()
+        self.order.order_line[0].qty_delivered = 3
+        self.assertEqual(self.order.order_line[0].delivery_state, "done")
+
+    def test_no_delivery_delivery_cost(self):
+        self._add_delivery_cost_line()
+        with self._mock_delivery():
+            self.assertEqual(self.order.order_line[1].delivery_state, "no")
+
+    def test_unprocessed_delivery_delivery_cost(self):
+        self._add_delivery_cost_line()
+        with self._mock_delivery():
+            self.order.action_confirm()
+            self.assertEqual(self.order.order_line[0].delivery_state, "unprocessed")
+            self.assertEqual(self.order.order_line[1].delivery_state, "no")
+
+    def test_partially_delivery_cost(self):
+        self._add_delivery_cost_line()
+        with self._mock_delivery():
+            self.order.action_confirm()
+            self.order.order_line[0].qty_delivered = 2
+            self.assertEqual(self.order.order_line[0].delivery_state, "partially")
+            self.assertEqual(self.order.order_line[1].delivery_state, "no")
+
+    def test_forced_delivery_cost(self):
+        self._add_delivery_cost_line()
+        with self._mock_delivery():
+            self.order.action_confirm()
+            self.order.order_line[0].qty_delivered = 2
+            self.order.order_line[0].force_delivery_state = True
```

Comment:
You can increase coverage a little bit by calling `self.order.order_line[0].action_force_delivery_state()` here, and then extend the test calling `self.order.order_line[0].action_unforce_delivery_state()` again (and checking the resulting delivery state).